ui/src/utils/lineWriter.ts

```diff
@@ -0,0 +1,129 @@
+import AJAX from 'src/utils/ajax'
+
+export interface Tags {
+  [key: string]: string
+}
+
+export interface Fields {
+  [key: string]: number | string
+}
+
+export enum Precision {
+  ns = 'ns',
+  u = 'u',
+  ms = 'ms',
+  s = 's',
+  m = 'm',
+  h = 'h',
+}
+
+const nowInSeconds = function nowInSeconds() {
+  return Math.floor(Date.now() / 1000)
+}
+
+// Build a line writer that can write arbitrary line data
+export const createLineFromModel = function createLineFromModel(
+  measurement: string,
+  fields: Fields,
+  tags: Tags = {},
+  timestamp: number = nowInSeconds()
+): string {
+  let tagString = ''
+  Object.keys(tags)
+    // Sort keys for a little extra perf
+    // https://v2.docs.influxdata.com/v2.0/write-data/best-practices/optimize-writes/#sort-tags-by-key
+    .sort()
+    .forEach((tagKey, i, tagKeys) => {
+      const tagValue = tags[tagKey]
+      const printableTagKey = tagKey
+        .replace(/\s+/g, '\\ ')
+        .replace(/,/g, '\\,')
+        .replace(/=/g, '\\=')
+
+      const printableTagValue = tagValue
+        .replace(/\n/g, '')
+        .replace(/\s+/g, '\\ ')
+        .replace(/,/g, '\\,')
+        .replace(/=/g, '\\=')
+
+      tagString = `${tagString}${printableTagKey}=${printableTagValue}`
+
+      // if this isn't the end of the string, append a comma
+      if (i < tagKeys.length - 1) {
+        tagString = `${tagString},`
+      }
+    })
+
+  let fieldString = ''
+  Object.keys(fields).forEach((fieldKey, i, fieldKeys) => {
+    const fieldValue = fields[fieldKey]
+
+    const printableFieldKey = fieldKey
+      .replace(/\s+/g, '\\ ')
+      .replace(/,/g, '\\,')
+      .replace(/=/g, '\\=')
+
+    let printableFieldValue = fieldValue
+    if (typeof fieldValue === 'string') {
+      printableFieldValue = fieldValue
+        .replace(/\n/g, '')
+        .replace(/\\/g, '\\')
+        .replace(/"/g, '\\"')
+    }
+
+    fieldString = `${fieldString}${printableFieldKey}=${printableFieldValue}`
+
+    // if this isn't the end of the string, append a comma
+    if (i < fieldKeys.length - 1) {
+      fieldString = `${fieldString},`
+    }
+  })
+
+  let lineStart = measurement.replace(/\s+/g, '\\ ').replace(/,/g, '\\,')
+  if (tagString !== '') {
+    lineStart = `${lineStart},${tagString}`
+  }
+
+  return `${lineStart} ${fieldString} ${timestamp}`
+}
+
+// Builds a basic line writer that will write a line to the url, org and bucket specified.
+// Returns a function named `writeLine` that utilizes closures to access the configuration
+// arguments passed in to `buildLineWriter`.
+export const buildLineWriter = function buildLineWriter(
```

Comment:
One thing we might want to consider it building up a buffer of metrics instead of just writing a single line. Maybe something on the order of ~100 metrics before making the actual network request.

Common patterns I've seen for influx writers are to have batch size and interval be configurable options. For example write 100 metrics in a batch or every 10s, whichever is first.

Reply:
@desa 
right on thanks! i figured that would be a suggestion :)

how does batching writes affect precision, if at all?

for that matter, what kind of precision is appropriate here? is second precision too coarse? could second-level precision lead to a situation where if there are two writes in the same second, the later one overwrites the earlier one?

Reply:
Typically each point in a batch shares a common timestamp precision. Second precision here is probably fine. I cant imagine we'd want to track these with sub second precision

Reply:
> could second-level precision lead to a situation where if there are two writes in the same second, the later one overwrites the earlier one?

Could happen, but I think if we're reporting with that level of frequency we're doing something wrong